bench/bench_spsc_queue.ml

```diff
@@ -1,46 +1,96 @@
-      Queue.pop_exn t |> ignore
-    done;
-    let n = Random.int ((1 lsl size_exponent) + 1) in
-    for i = 1 to n do
-      Queue.push_exn t i
-    done
-  in
+let run_one ~unsafe ~budgetf ?(size_exponent = 3)
+    ?(n_msgs = 80 * Util.iter_factor) () =
   let init _ = () in
-  let work i () =
-    if i = 0 then
-      let rec loop n =
-        if 0 < n then
-          if Queue.try_push t n then loop (n - 1)
-          else begin
-            Domain.cpu_relax ();
-            loop n
-          end
+  let work, before =
+    if unsafe then
+      let module Queue = Saturn_lockfree.Single_prod_single_cons_queue_unsafe in
+      let t = Queue.create ~size_exponent in
+
+      let before () =
+        while Queue.size t <> 0 do
+          Queue.pop_exn t |> ignore
+        done;
+        let n = Random.int ((1 lsl size_exponent) + 1) in
+        for i = 1 to n do
+          Queue.push_exn t i
+        done
+      in
+      let work i () =
+        if i = 0 then
+          let rec loop n =
+            if 0 < n then
+              if Queue.try_push t n then loop (n - 1)
+              else begin
+                Domain.cpu_relax ();
+                loop n
+              end
+          in
+          loop n_msgs
+        else
+          let rec loop n =
+            if 0 < n then
+              match Queue.pop_opt t with
+              | Some _ -> loop (n - 1)
+              | None ->
+                  Domain.cpu_relax ();
+                  loop n
+          in
+          loop n_msgs
       in
-      loop n_msgs
+      (work, before)
     else
-      let rec loop n =
-        if 0 < n then
-          match Queue.pop_opt t with
-          | Some _ -> loop (n - 1)
-          | None ->
-              Domain.cpu_relax ();
-              loop n
+      let module Queue = Saturn_lockfree.Single_prod_single_cons_queue in
+      let t = Queue.create ~size_exponent in
+
+      let before () =
+        while Queue.size t <> 0 do
+          Queue.pop_exn t |> ignore
+        done;
+        let n = Random.int ((1 lsl size_exponent) + 1) in
+        for i = 1 to n do
+          Queue.push_exn t i
+        done
+      in
+      let work i () =
+        if i = 0 then
+          let rec loop n =
+            if 0 < n then
+              if Queue.try_push t n then loop (n - 1)
+              else begin
+                Domain.cpu_relax ();
+                loop n
+              end
+          in
+          loop n_msgs
+        else
+          let rec loop n =
+            if 0 < n then
+              match Queue.pop_opt t with
+              | Some _ -> loop (n - 1)
+              | None ->
+                  Domain.cpu_relax ();
+                  loop n
+          in
+          loop n_msgs
       in
-      loop n_msgs
+      (work, before)
   in
 
-  let config = Printf.sprintf "2 workers, capacity %d" (1 lsl size_exponent) in
+  let config =
+    Printf.sprintf "2 workers, capacity %d (%s)" (1 lsl size_exponent)
+      (if unsafe then "unsafe" else "safe")
```

Comment:
Hmm... So, this will add `(safe)` or `(unsafe)` as a suffix.

How about just adding the `(unsafe)` suffix and not adding a suffix for "safe" versions?

This would be nice in two ways:
1. The existing benchmark does not have the `(safe)` suffix.  So, keeping the existing naming means that there is no discontinuity in the benchmark graphs.
2. If, at some point, OCaml is enhanced to avoid the need for the unsafe versions, we can then just drop the unsafe version and don't have to rename the safe versions, because at that point a `(safe)` suffix would be superfluous.

Reply:
Good point, thanks !